ndk/src/media/media_extractor.rs

```diff
@@ -0,0 +1,651 @@
+        MediaError::from_status(status).unwrap()
+    }
+}
+
+/// List of crypto schemes and their data.
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct PsshInfo<'a>(ffi::PsshInfo, PhantomData<&'a PsshEntry<'a>>);
+
+impl<'a> PsshInfo<'a> {
+    pub fn entries(&self) -> &[ffi::PsshEntry] {
+        unsafe { self.0.entries.as_slice(self.0.numentries) }
+    }
+}
+
+/// Mapping of crypto scheme uuid to the scheme specific data for that scheme.
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct PsshEntry<'a>(ffi::PsshEntry, PhantomData<&'a [u8]>);
+
+impl<'a> PsshEntry<'a> {
+    pub fn uuid(&self) -> ffi::AMediaUUID {
+        self.0.uuid
+    }
+
+    pub fn data(&self) -> &[u8] {
+        unsafe { std::slice::from_raw_parts(self.0.data.cast(), self.0.datalen) }
+    }
+}
+
+/// Called to request data from the given `offset`.
+///
+/// Implementations should should write up to the size of the slice number of bytes into `buffer`,
+/// and return the number of bytes written.
+///
+/// Return `0` if the slice is empty (thus no bytes are read).
+///
+/// Return [`None`] to indicate that end of stream is reached.
+#[cfg(feature = "api-level-28")]
+#[doc(alias = "AMediaDataSourceReadAt")]
+pub type MediaDataSourceReadAt = Box<dyn FnMut(usize, &mut [u8]) -> Option<usize> + Send + Sync>;
+
+/// Called to get the size of the data source.
+///
+/// Return the size of data source in bytes, or [`None`] if the size is unknown.
+#[cfg(feature = "api-level-28")]
+#[doc(alias = "AMediaDataSourceGetSize")]
+pub type MediaDataSourceGetSize = Box<dyn FnMut() -> Option<usize> + Send + Sync>;
+
+/// Called to close the data source, unblock reads, and release associated resources.
+///
+/// The NDK media framework guarantees that after the first [`MediaDataSourceCallbacks::close`]
+/// is called, no future callbacks will be invoked on the data source except for
+/// [`MediaDataSourceCallbacks::close`] itself.
+///
+/// Closing a data source allows [`MediaDataSourceCallbacks::read_at`] calls that were blocked
+/// waiting for I/O data to return promptly.
+///
+/// When using [`MediaDataSource`] as input to [`MediaExtractor`], closing has the
+/// effect of unblocking slow reads inside of [`MediaExtractor::set_data_source()`] and
+/// [`MediaExtractor::read_sample_data()`].
+#[cfg(feature = "api-level-28")]
+#[doc(alias = "AMediaDataSourceClose")]
+pub type MediaDataSourceClose = Box<dyn FnMut() + Send + Sync>;
+
+/// Called to get an estimate of the number of bytes that can be read from this data source starting
+/// at `offset` without blocking for I/O.
+///
+/// Return [`None`] when such an estimate is not possible.
+#[cfg(feature = "api-level-29")]
+#[doc(alias = "AMediaDataSourceGetAvailableSize")]
+pub type MediaDataSourceGetAvailableSize =
+    Box<dyn FnMut(Option<usize>) -> Option<usize> + Send + Sync>;
+
+/// Callbacks for [`MediaDataSource`].
+#[derive(Default)]
+#[cfg(feature = "api-level-28")]
+pub struct MediaDataSourceCallbacks {
```

Comment:
Feels a bit unidiomatic. It C these context/user data pointers are extremely common in callback contexts, but this is really just a workaround for the fact C doesn't have closures. I'm not sure I'd propagate this to the Rust API unless there is a particularly compelling reason to do so.

Reply:
@spencercw the main reason is having a single place to own a `Send+Sync` object that is made available to all the callbacks, instead of being required to `Copy`/`Clone` it into every callback that the user wishes to set.

Reply:
That makes sense, but it's a small optimisation and I'm not sure it's worth the additional API complication. In all likelihood the shared data is probably going to be something cheap to copy like an `Arc` or channel anyway. I don't feel strongly either way though.

Reply:
It doesn't seem very complicated but let's leave it then. We shouldn't emit mutable borrows so users are likely already required to wrap in some other structure anyway.